File: applicatie/src/main/java/com/vdzon/administratie/pdfgenerator/overzicht/BoekingsVerrijker.java
package com.vdzon.administratie.pdfgenerator.overzicht;

import com.vdzon.administratie.model.Administratie;
import com.vdzon.administratie.model.Afschrift;
import com.vdzon.administratie.model.Factuur;
import com.vdzon.administratie.model.Rekening;
import com.vdzon.administratie.model.boekingen.*;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetAfschrift;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetRekening;
import com.vdzon.administratie.mongo.BigDecimalConverter;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.List;
import java.util.stream.Collectors;

public class BoekingsVerrijker {
    public static List<VerrijkteBoeking> verrijk(Administratie administratie){
        return administratie.getBoekingen()
                .stream()
                .map(boeking->verrijkBoeking(boeking, administratie))
                .collect(Collectors.toList());
    }

    private static VerrijkteBoeking verrijkBoeking(Boeking boeking, Administratie administratie) {
        if (boeking instanceof InkomstenZonderFactuurBoeking){
            return verrijk((InkomstenZonderFactuurBoeking)boeking, administratie);
        }
        if (boeking instanceof BetalingZonderFactuurBoeking){
            return verrijk((BetalingZonderFactuurBoeking)boeking, administratie);
        }
        if (boeking instanceof PriveBetalingBoeking){
            return verrijk((PriveBetalingBoeking)boeking, administratie);
        }
        if (boeking instanceof BetaaldeFactuurBoeking){
            return verrijk((BetaaldeFactuurBoeking)boeking, administratie);
        }
        if (boeking instanceof BetaaldeRekeningBoeking){
            return verrijk((BetaaldeRekeningBoeking)boeking, administratie);
        }

        throw new RuntimeException("Boeking is van onbekend type : "+boeking.getClass().getCanonicalName());
    }


    private static VerrijkteBoeking verrijk(InkomstenZonderFactuurBoeking boeking, Administratie administratie) {
        Afschrift afschift = getAfschift(boeking, administratie);
        VerrijkteBoeking.BOEKINGSTYPE boekingsType = VerrijkteBoeking.BOEKINGSTYPE.INKOMSTEN_ZONDER_FACTUUR;
        BigDecimal boekingsBedrag = afschift.getBedrag();

        return VerrijkteBoeking.newBuilder()
                .afschrift(afschift)
                .afschriftBedrag(afschift.getBedrag())
                .afschriftDate(afschift.getBoekdatum())
                .boeking(boeking)
                .boekingsBedrag(boekingsBedrag)
                .boekingsType(boekingsType)
                .build();
    }

    private static VerrijkteBoeking verrijk(BetalingZonderFactuurBoeking boeking, Administratie administratie) {
        Afschrift afschift = getAfschift(boeking, administratie);
        VerrijkteBoeking.BOEKINGSTYPE boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETALING_ZONDER_FACTUUR;
        BigDecimal boekingsBedrag = afschift.getBedrag();

        return VerrijkteBoeking.newBuilder()
                .afschrift(afschift)
                .afschriftBedrag(afschift.getBedrag())
                .afschriftDate(afschift.getBoekdatum())
                .boeking(boeking)
                .boekingsBedrag(boekingsBedrag)
                .boekingsType(boekingsType)
                .build();
    }

    private static VerrijkteBoeking verrijk(PriveBetalingBoeking boeking, Administratie administratie) {
        Afschrift afschift = getAfschift(boeking, administratie);
        VerrijkteBoeking.BOEKINGSTYPE boekingsType = VerrijkteBoeking.BOEKINGSTYPE.PRIVE_BETALING;
        BigDecimal boekingsBedrag = afschift.getBedrag();

        return VerrijkteBoeking.newBuilder()
                .afschrift(afschift)
                .afschriftBedrag(afschift.getBedrag())
                .afschriftDate(afschift.getBoekdatum())
                .boeking(boeking)
                .boekingsBedrag(boekingsBedrag)
                .boekingsType(boekingsType)
                .build();
    }

    private static VerrijkteBoeking verrijk(BetaaldeFactuurBoeking boeking, Administratie administratie) {
        Afschrift afschift = getAfschift(boeking, administratie);
        Factuur factuur = getFactuur(boeking, administratie);
        if (factuur == null){
            factuur = Factuur.newBuilder().factuurDate(afschift.getBoekdatum()).build();
        }
        VerrijkteBoeking.BOEKINGSTYPE boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETAALDE_FACTUUR;
        BigDecimal boekingsBedrag = afschift.getBedrag();
        BigDecimal factuurBedrag = factuur.getBedragIncBtw();
        LocalDate factuurDate = factuur.getFactuurDate();

        return VerrijkteBoeking.newBuilder()
                .afschrift(afschift)
                .afschriftBedrag(afschift.getBedrag())
                .afschriftDate(afschift.getBoekdatum())
                .boeking(boeking)
                .boekingsBedrag(boekingsBedrag)
                .boekingsType(boekingsType)
                .factuur(factuur)
                .factuurBedrag(factuurBedrag)
                .factuurDate(factuurDate)
                .build();
    }

    private static VerrijkteBoeking verrijk(BetaaldeRekeningBoeking boeking, Administratie administratie) {
        Afschrift afschift = getAfschift(boeking, administratie);
        Rekening rekening = getRekening(boeking, administratie);
        VerrijkteBoeking.BOEKINGSTYPE boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETAALDE_REKENING;
        BigDecimal boekingsBedrag = afschift.getBedrag();

        return VerrijkteBoeking.newBuilder()
                .afschrift(afschift)
                .afschriftBedrag(afschift.getBedrag())
                .afschriftDate(afschift.getBoekdatum())
                .boeking(boeking)
                .boekingsBedrag(boekingsBedrag)
                .boekingsType(boekingsType)
                .rekening(rekening)
                .rekeningBedrag(rekening.getBedragIncBtw())
                .rekeningDate(rekening.getRekeningDate())
                .build();
    }

    private static Rekening getRekening(BoekingMetRekening boeking, Administratie administratie) {
        return administratie.getRekeningen()
                .stream()
                .filter(rekening -> rekening.getRekeningNummer().equals(boeking.getRekeningNummer()))
                .findFirst()
                .orElse(null);
    }

    private static Factuur getFactuur(BoekingMetFactuur boeking, Administratie administratie) {
        return administratie.getFacturen()
                .stream()
                .filter(factuur -> factuur.getFactuurNummer().equals(boeking.getFactuurNummer()))
                .findFirst()
                .orElse(null);
    }

    private static Afschrift getAfschift(BoekingMetAfschrift boeking, Administratie administratie) {
        return administratie.getAfschriften()
                .stream()
                .filter(afschrift -> afschrift.getNummer().equals(boeking.getAfschriftNummer()))
                .findFirst()
                .orElse(null);
    }
}

File: applicatie/src/main/java/com/vdzon/administratie/pdfgenerator/overzicht/BoekingsVerrijker.kt
package com.vdzon.administratie.pdfgenerator.overzicht

import com.vdzon.administratie.model.*
import com.vdzon.administratie.model.boekingen.*
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetAfschrift
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetRekening

import java.math.BigDecimal
import java.time.LocalDate
import java.util.ArrayList
import java.util.stream.Collectors

object BoekingsVerrijker {
    fun verrijk(administratie: Administratie): List<VerrijkteBoeking> {
        return administratie.boekingen
                .map{ boeking -> verrijkBoeking(boeking, administratie) }
    }

    private fun verrijkBoeking(boeking: Boeking, administratie: Administratie): VerrijkteBoeking {
        if (boeking is InkomstenZonderFactuurBoeking) {
            return verrijk(boeking, administratie)
        }
        if (boeking is BetalingZonderFactuurBoeking) {
            return verrijk(boeking, administratie)
        }
        if (boeking is PriveBetalingBoeking) {
            return verrijk(boeking, administratie)
        }
        if (boeking is BetaaldeFactuurBoeking) {
            return verrijk(boeking, administratie)
        }
        if (boeking is BetaaldeRekeningBoeking) {
            return verrijk(boeking, administratie)
        }

        throw RuntimeException("Boeking is van onbekend type : " + boeking.javaClass.canonicalName)
    }


    private fun verrijk(boeking: InkomstenZonderFactuurBoeking, administratie: Administratie): VerrijkteBoeking {
        val afschift = getAfschift(boeking, administratie)
        val boekingsType = VerrijkteBoeking.BOEKINGSTYPE.INKOMSTEN_ZONDER_FACTUUR
        val boekingsBedrag = afschift?.bedrag?: BigDecimal.ZERO

        return VerrijkteBoeking(
                boeking,
                null,
                null,
                afschift,
                boekingsType,
                boekingsBedrag,
                afschift?.bedrag,
                BigDecimal.ZERO,
                BigDecimal.ZERO,
                afschift?.boekdatum,
                null,
                null)
    }

    private fun verrijk(boeking: BetalingZonderFactuurBoeking, administratie: Administratie): VerrijkteBoeking {
        val afschift = getAfschift(boeking, administratie)
        val boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETALING_ZONDER_FACTUUR
        val boekingsBedrag = afschift?.bedrag?: BigDecimal.ZERO

        return VerrijkteBoeking(
                boeking,
                null,
                null,
                afschift,
                boekingsType,
                boekingsBedrag,
                afschift?.bedrag,
                BigDecimal.ZERO,
                BigDecimal.ZERO,
                afschift?.boekdatum,
                null,
                null)
    }

    private fun verrijk(boeking: PriveBetalingBoeking, administratie: Administratie): VerrijkteBoeking {
        val afschift = getAfschift(boeking, administratie)
        val boekingsType = VerrijkteBoeking.BOEKINGSTYPE.PRIVE_BETALING
        val boekingsBedrag = afschift?.bedrag?: BigDecimal.ZERO

        return VerrijkteBoeking(
                boeking,
                null,
                null,
                afschift,
                boekingsType,
                boekingsBedrag,
                afschift?.bedrag,
                BigDecimal.ZERO,
                BigDecimal.ZERO,
                afschift?.boekdatum,
                null,
                null)
    }

    private fun verrijk(boeking: BetaaldeFactuurBoeking, administratie: Administratie): VerrijkteBoeking {
        val afschift = getAfschift(boeking, administratie)
        var factuur: Factuur? = getFactuur(boeking, administratie)
        if (factuur == null) {
            factuur = Factuur(
                    "",
                    "",
                    "",
                    afschift?.boekdatum,
                    null,
                    ArrayList<FactuurRegel>(),
                    BigDecimal.ZERO,
                    BigDecimal.ZERO,
                    BigDecimal.ZERO)
        }
        val boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETAALDE_FACTUUR
        val boekingsBedrag = afschift?.bedrag?: BigDecimal.ZERO
        val factuurBedrag = factuur.bedragIncBtw
        val factuurDate = factuur.factuurDate

        return VerrijkteBoeking(
                boeking,
                factuur,
                null,
                afschift,
                boekingsType,
                boekingsBedrag,
                afschift?.bedrag,
                factuurBedrag,
                BigDecimal.ZERO,
                afschift?.boekdatum,
                factuurDate,
                null)

    }

    private fun verrijk(boeking: BetaaldeRekeningBoeking, administratie: Administratie): VerrijkteBoeking {
        val afschift = getAfschift(boeking, administratie)
        val rekening = getRekening(boeking, administratie)
        val boekingsType = VerrijkteBoeking.BOEKINGSTYPE.BETAALDE_REKENING
        val boekingsBedrag = afschift?.bedrag?: BigDecimal.ZERO

        return VerrijkteBoeking(
                boeking,
                null,
                rekening,
                afschift,
                boekingsType,
                boekingsBedrag,
                afschift?.bedrag,
                BigDecimal.ZERO,
                rekening?.bedragIncBtw,
                afschift?.boekdatum,
                null,
                rekening?.rekeningDate)
    }

    private fun getRekening(boeking: BoekingMetRekening, administratie: Administratie): Rekening? {
        return administratie.rekeningen
                .filter{ rekening -> rekening.rekeningNummer == boeking.rekeningNummer }
                .firstOrNull()
    }

    private fun getFactuur(boeking: BoekingMetFactuur, administratie: Administratie): Factuur? {
        return administratie.facturen
                .filter{ factuur -> factuur.factuurNummer == boeking.factuurNummer }
                .firstOrNull()
    }

    private fun getAfschift(boeking: BoekingMetAfschrift, administratie: Administratie): Afschrift? {
        return administratie.afschriften
                .filter{ afschrift -> afschrift.nummer == boeking.afschriftNummer }
                .firstOrNull()
    }
}
